scipy/stats/_sampling.py

```diff
@@ -0,0 +1,205 @@
+import numpy as np
+from scipy._lib._util import check_random_state
+
+
+class RatioUniforms:
+    """
+    Generate random samples from a probability density function using the
+    ratio-of-uniforms method.
+
+    Parameters
+    ----------
+    pdf : callable
+        A function with signature `pdf(x)` that is proportional to the
+        probability density function of the distribution.
```

Comment:
Ok this is the old API, but do we maybe now want to allow passing a class which has a `pdf` method? That would allow passing any distribution object, including the new things we are building.

Reply:
I left this one out for now: one can simply instantiate the class with "pdf=dist.pdf" if needed (for frozen distributions). allowing distribution objects was a simplification for other methods like `NumericalInversePolynomial` since you can extract the pdf, cdf, logpdf, support etc from the object whereas here it is really just the pdf, so passing distribution objects might be an overkill. it can be added in a follow-up w/o breaking anything. let me know what you think

Reply:
@mdhaber Can you weight in? My argument is that we typically would have a class with a pdf method, not just a pdf method as is. So it seems natural to me to require a class with a pdf method. Otherwise any normal use of that from a SciPy's perspective would need the user to pass the method only.

Reply:
I would prefer to keep it as `pdf : callable`. This enables it to be used as a building block outside of the `scipy.stats` distribution framework.

Reply:
It would still be usable from outside:

```python
class MyPDF:
    def pdf(x):
        ...
```

The question to me is if we want to have a better integration with our infra or not. If we say that `rvs` is a good name because of existing infra, then the argument is the same here as all other classes from this submodules take a class as input. Changing that would make it an exception.

Reply:
If it were difficult to get the PDF callable from a distribution object, then yes, a `pdf : callable` argument would cause a problem for integration with the rest of our infrastructure. But it's just a matter of accessing the `.pdf` method, a very prominent part of the distribution infrastructure's public API. That's just one of the things you are supposed to do with distribution objects. A function that takes a `pdf : callable` argument is no less integrated with our infrastructure than one that takes a `dist : rv_generic`.

Reply:
Just to be clear with what I am proposing. I am proposing a generic class, it does not have to be `rv_generic`. It only needs to have a `pdf` method.

Reply:
That's even more inconsistent with the rest of the infrastructure, then.

Reply:
How so? That's exactly the behavior of the other classes.

Anyway, not going to die on this hill

Reply:
All of the other classes are `rv_generic`s with the full suite of methods. "class with a `pdf` method" is a new concept to the infrastructure that would be unique to this function. If we wanted to allow your workaround, we'd have to document it that way, create the new type annotation for it, etc. just for this one function.

Reply:
Fair, I retract the claim that it's more inconsistent.

Reply:
@tupui let me know if you think this should be changed in a follow-up. We still have a lot of time until the release. I agree it differs slightly from the other methods. I thought it is ok since it is a bit easier to use (either pass `dist.pdf` or call it with `pdf= lambda x: ...`) and I did not mean to achieve 100% consistency as this is a separate Python implementation and it does not rely on some of the generic methods that Tirth implemented for the UNU.RAN methods (e.g., the support of frozen distributions relies on a helper function to call _pdf taking into account the support (to ensure that pdf is zero outside of it)).